feat(core): add new infrastructure search tagging widget

icfantv opened the pull request:
This is currently feature flagged behind `INF_SEARCH_ENABLED` and is currently set to `false`.  I wanted to get this in before the PR got too big for people to actually review (or want to review).  The searching is not wired in yet as the backend work still needs to be determined.
![screen shot 2017-07-29 at 10 47 23 pm](https://user-images.githubusercontent.com/370054/28751077-73bb7690-74b2-11e7-80be-fc65bd834445.png)

jrsquared commented on app/scripts/modules/core/src/search/widgets/Search.tsx:
```diff
@@ -0,0 +1,365 @@
+
+  private handleBlur(): void {
+    this.setFocusState(false);
+    this.setState({
```
Historically not a good pattern for call setState multiple times in a row; can cause weird UI flickering and such.

jrsquared commented on app/scripts/modules/core/src/search/widgets/Search.tsx:
```diff
@@ -0,0 +1,365 @@
+        isOpen: true
+      });
+    } else {
+      this.setState({
```
Not a big deal, just a pattern that I like is to have one `this.setState()` call in the function so it's easier to grok that there really is only one setState happening.

jrsquared commented on app/scripts/modules/core/src/search/widgets/Search.tsx:
```diff
@@ -0,0 +1,365 @@
+  }
+
+  private handleFocus(): void {
+    this.setFocusState(true);
```
Maybe you don't even need `this.setFocusState(focus)` since it's really just wrapping `this.setState({focused: focus})` Doesn't really save much, and then causes double setState calls

jrsquared commented on app/scripts/modules/core/src/search/widgets/Search.tsx:
```diff
@@ -0,0 +1,365 @@
+        });
+      }
+    } else {
+      this.setState({
```
Same pattern, one setState call.

jrsquared commented on app/scripts/modules/core/src/search/widgets/search.component.ts:
```diff
@@ -0,0 +1,12 @@
+import {module} from 'angular';
+import {react2angular} from 'react2angular';
```
give those brackets room to breathe

jrsquared commented on app/scripts/modules/core/src/widgets/tags/Tag.spec.tsx:
```diff
@@ -0,0 +1,102 @@
+
+  let component: ReactWrapper<ITagProps, any>;
+
+  function noop() {}
```
kill it with fire

jrsquared commented on app/scripts/modules/core/src/widgets/tags/Tag.spec.tsx:
```diff
@@ -0,0 +1,102 @@
+
+  function noop() {}
+
+  function getNewTag(): ITag {
```
what is this `function` word?

jrsquared commented on app/scripts/modules/core/src/widgets/tags/Tag.tsx:
```diff
@@ -0,0 +1,83 @@
+@autoBindMethods
+export class Tag extends React.Component<ITagProps> {
+
+  constructor(props: ITagProps) {
```
kill the constructor

jrsquared commented on app/scripts/modules/core/src/widgets/tags/TagList.spec.tsx:
```diff
@@ -0,0 +1,38 @@
+
+  let component: ReactWrapper<ITagListProps, any>;
+
+  function noop() {}
```
kill it

jrsquared commented on app/scripts/modules/core/src/widgets/tags/TagList.spec.tsx:
```diff
@@ -0,0 +1,38 @@
+    return mount(
+      <TagList
+        tags={[1, 2, 3].map((seed: number) => getNewTag(seed))}
+        onBlur={noop}
```
make all these props optional

anotherchrisberry commented on app/scripts/modules/core/src/search/widgets/Filter.tsx:
```diff
@@ -0,0 +1,44 @@
+        onKeyUp={this.handleKeyUp}
+      >
+        <div className="filter__text">{text}</div>
+        <div className="filter__modifier">[{modifier.toLocaleUpperCase()}:]</div>
```
also, why the double anything? We don't do that anywhere else in the codebase with class names.

icfantv replied:
According to our [conventions document](https://github.com/spinnaker/deck/wiki/Conventions), we use BEM for class names so I was just following that.  I do want to be consistent but I feel like I'm getting mixed messages when I'm told "it's in the coding conventions doc" and when I follow it I also get questions.  Is that not fair?  If the rule of thumb is "Follow the coding conventions document unless the code dictates otherwise," I'm not sure there's value in having coding conventions.  What say ye?

anotherchrisberry replied:
I didn't realize that was coming out of BEM.

The BEM entry was more aspirational than anything, as we really haven't followed it at all since I put that in the conventions document a year ago.

The rule of thumb is to follow the coding conventions, but if the actual code conflicts with the conventions, we should revise the conventions so they're aligned with how we're writing code, or change how we write code.

I like the idea of BEM and would be happy to start using it more frequently, like here, but will defer to others.

anotherchrisberry commented on app/scripts/modules/core/src/search/widgets/Filters.tsx:
```diff
@@ -0,0 +1,83 @@
+import * as classNames from 'classnames';
+import autoBindMethods from 'class-autobind-decorator';
+
+import { Key } from '@spinnaker/core';
```
Within `core`, imports should be relative or with `core` as the root, e.g. `core/widgets/keys`

anotherchrisberry commented on app/scripts/modules/core/src/search/widgets/Search.tsx:
```diff
@@ -0,0 +1,365 @@
+import * as classNames from 'classnames';
+import autoBindMethods from 'class-autobind-decorator';
+
+import { IQueryParams, ITag, Key, TagList, UrlParser } from '@spinnaker/core';
```
already in core

anotherchrisberry commented on app/scripts/modules/core/src/search/widgets/Search.tsx:
```diff
@@ -0,0 +1,365 @@
+
+    let result = input;
+    if (this.hasModifier(input)) {
+      result = input.substring(input.indexOf(':') + 1, input.length);
```
`result = input.split(':')[1]` is a little easier to understand IMO

anotherchrisberry commented on app/scripts/modules/core/src/search/widgets/Search.tsx:
```diff
@@ -0,0 +1,365 @@
+      result = input.substring(input.indexOf(':') + 1, input.length);
+    }
+
+    return result;
```
The whole method could be:
```javascript
return this.hasModifier(input) ? input.split(':')[1]: input;
```

anotherchrisberry commented on app/scripts/modules/core/src/search/widgets/Search.tsx:
```diff
@@ -0,0 +1,365 @@
+      const activeIndex = this.getActiveFilterIndex();
+      if (activeIndex === (filterTypes.length - 1)) {
+        const activeFilter = filterTypes[0];
+        this.inputElement.value = this.getActiveFilterText(activeFilter);
```
feels like setting the value should be the responsibility of the `render` method, rather than something done within an event handler, but will defer to @jrsquared on that

jrsquared replied:
Oh yeah, good catch. This should definitely just be set in the render function. This event handle should just call `setState` with a new value, and `value` should be set to the state in the render function.

anotherchrisberry commented on app/scripts/modules/core/src/search/widgets/SearchFilterTypeRegistry.ts:
```diff
@@ -0,0 +1,47 @@
+    let result = SearchFilterTypeRegistry.registry.get(key);
+    if (key === SearchFilterTypeRegistry.NAME_FILTER.key) {
+      result = SearchFilterTypeRegistry.NAME_FILTER;
+    }
```
Is there a reason to not put keyword and name into the registry? Also, any reason to exclude keyword checking here?

anotherchrisberry commented on app/scripts/modules/core/src/domain/Keys.ts:
```diff
@@ -0,0 +1,34 @@
+  static get UP_ARROW(): string {
+    return 'ArrowUp';
+  }
+}
```
Maybe put these in the `react` module?

icfantv commented on app/scripts/modules/core/src/search/widgets/Filter.tsx:
```diff
@@ -0,0 +1,44 @@
+export interface IFilterProps {
+  filterType: IFilterType;
+  isActive: boolean;
+  onKeyUp(event: React.KeyboardEvent<HTMLElement>): void;
```
Yikes, that was NOT what that was supposed to look like.  It should have been `onKeyUp: (event: React.KeyboardEvent<HTMLEVent>) => void`.  I didn't even realize this was valid syntax...

christopherthielen commented on app/scripts/modules/core/src/search/widgets/Search.tsx:
```diff
@@ -0,0 +1,365 @@
+    }
+  }
+
+  private navigateDown(): void {
```
Seems like navigateUp and navigateDown could be collapsed into a single method

christopherthielen commented on app/scripts/modules/core/src/search/widgets/SearchFilterTypeRegistry.ts:
```diff
@@ -0,0 +1,47 @@
+
+export class SearchFilterTypeRegistry {
+
+  public static KEYWORD_FILTER: IFilterType = Object.freeze({
```
What is the reasoning behind freezing these?

anotherchrisberry commented on app/scripts/modules/amazon/src/aws.module.ts:
```diff
@@ -114,6 +114,9 @@ module(AMAZON_MODULE, [
   });
+}).run(() => {
+  SearchFilterTypeRegistry.register({key: 'account', modifier: 'acct', text: 'Account'});
+  SearchFilterTypeRegistry.register({key: 'region', modifier: 'reg', text: 'Region'});
```
Nice, I like that this is pluggable.

anotherchrisberry commented:
Finally pulled this down and played around with it. Looks nice so far. Some minor UX things:
* both search fields (old and new) are showing
* when I have some tags in place and type the delete key, it doesn't remove the tag (although I can use the arrow key to highlight the tag, then press delete to remove it)
* when the search field is empty and I press the down arrow, it pre-selects "Key", but doesn't add "key:" to the search field; if I press down again, it adds "name:". Seems like it should go ahead and add "key:" on that first arrow down.
* if I press escape when a filter category is highlighted, the menu collapses, but the category remains in the search field; seems like it should clear the search field
* when some tags are populated and I type the right arrow key, I am surprised that the cursor ends up on the first tag
* if I type a field key, it auto-highlights it from the menu, but if I then enter a space and a value, e.g. "reg: us-east-1", pressing enter does not add it as a tag. Should probably allow that when entering tags
* when I see "Keyword" and "key:", I think that's a specific field, maybe a tag on an object. Should we rename that the label for "key" to "Any field", maybe?
* it would be nice to capture "ctrl+a" when the input has the cursor and be able to delete all the tags already there - same with holding the shift key to select multiple tags and be able to delete them, although that definitely seems like something we could add later

It would be cool (but, again, later) to pre-populate some of these fields, specifically account and region, so when the user started typing, e.g. "us-", we could limit the options displayed based on the known values.

icfantv commented on app/scripts/modules/core/src/search/widgets/Filters.tsx:
```diff
@@ -0,0 +1,83 @@
+    );
+  }
+
+  private generateFilterLayout(layout: IFiltersLayout): JSX.Element {
```
I agree.  In this case, it's not actually rendering, though, it is just generating the JSX.  Do you disagree?

christopherthielen replied:
I think the term `render` is well enough understood in react-land as "return an element", since that is what the full `render()` function does.

icfantv commented:
Updated per nearly all of the comments.
* Left `Filters` event handler as non-optional as it didn't seem to make sense, functionally, to do so.
* Implemented some of @anotherchrisberry's suggestions and captured the rest in a JIRA as some are more involved and I wanted to keep this PR small-ish.  Implemented:
  * 86'd old search field (note that search won't work until backend is done)
  * Adds the `key:` prefix for when menu is closed and user hits up/down arrow keys.
  * Escape will clear the input field if no search text is entered (i.e., only a modifier has been entered)

PTAL and LMK

anotherchrisberry commented on app/scripts/modules/core/src/search/widgets/Filters.spec.tsx:
```diff
@@ -0,0 +1,59 @@
+        activeFilter={activeFilter}
+        layouts={[1, 2].map((seed: number) => getNewLayout(seed))}
+        isOpen={isOpen}
+        onFilterChange={noop}
```
Looks like this is the only outstanding change

icfantv commented:
Renamed method and made event handler `Filters` optional.  Should be ready for merging.

anotherchrisberry commented:
merge away before it gets stale!
